Adjust UTxO set when building publish script txs

ffakenz opened the pull request:
Currently, Hydra needs to build 3 txs to publish scripts. Each is built using the faucet UTxO.  

This logic was adjusting the first selected UTxO repeatedly, causing the following [CI failure](https://github.com/cardano-scaling/hydra/actions/runs/14612733559/job/41012379950):
```
hydra-cluster: Illegal Value in TxOut: MaryValue (Coin (-64520700)) (MultiAsset (fromList []))
```

Now:
- we adjust the entire UTxO set on every tx build iteration, and then
- we select all outputs associated with the key address (instead of just the first matching one).

<!-- Describe your change here -->

---

<!-- Consider each and tick it off one way or the other -->
* [ ] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
adjust full utxo while building each publish script tx

* not just the first one selected to be spend, over and over

ffakenz committed:
pick all outputs by address

ch1bo commented on hydra-node/src/Hydra/Chain/ScriptRegistry.hs:
```diff
@@ -151,7 +150,7 @@ buildScriptPublishingTx pparams systemStart networkId eraHistory stakePools chan
         Right rawTx -> do
           let body = getTxBody rawTx
-          pure (makeSignedTransaction [makeShelleyKeyWitness body (WitnessPaymentKey sk)] body, body, utxoToSpend)
+          pure $ makeSignedTransaction [makeShelleyKeyWitness body (WitnessPaymentKey sk)] body
```
Returning `utxoToSpend` was the bug!

This is better now as we correctly chain the transactions by using the `tx` to adjust our available UTxO in line 121

ch1bo commented on hydra-node/src/Hydra/Chain/ScriptRegistry.hs:
```diff
@@ -118,11 +117,11 @@ buildScriptPublishingTxs pparams systemStart networkId eraHistory stakePools sta
       pure tx
  where
-  pickKeyAddressUTxO utxo = maybe mempty UTxO.singleton $ UTxO.findBy (\(_, txOut) -> isKeyAddress (txOutAddress txOut)) utxo
+  pickKeyAddressUTxO = UTxO.filter (isKeyAddress . txOutAddress)
```
This changes the code to spend _all_ UTxO owned by `sk`. I think this would work, but I'm not sure if this is what I would _always_ want to happen.

For example, I did use my personal signing key to publish scripts in the past. However, I don't want all my UTxO to be squashed just because I am publishing scripts with `hydra-node`.

ch1bo replied:
This is not true. It still only selects a single utxo (probably in the function further below).

ch1bo reviewed (commented):
Good catch! I'm not approving because this also changes the behavior on "coin selection".

In general, I feel this code is still way too complicated. The algorithm can be greatly simplified by just taking `txOuts' !! 1` as the next `utxoToSpend`. Also `StateT` is not really benefiting here and a simple loop would have done the trick too.

ch1bo committed:
Add a dedicated test of publishHydraScripts

ch1bo reviewed (approved):
I was wrong. The current code seems not to squash UTxO as proven by this test I added:

https://github.com/cardano-scaling/hydra/blob/701f82b0e5a5c64d35048ccbffc6ce207c1e52c0/hydra-cluster/test/Test/Hydra/Cluster/FaucetSpec.hs#L65-L77

ch1bo committed:
Simplify script publishing algorithm

It now only finds a suitable UTxO in the beginning and then just
respends the change output to create all reference script outputs.

ch1bo commented:
I have pushed a test and refactored the `buildScriptPublishingTxs` in that branch to have a simpler algorithm. With too little funds available it fails now with:

```
  1) Test.Hydra.Cluster.Faucet.publishHydraScriptsAs selects a suitable output
       uncaught exception: PublishScriptException
       FailedToFindSuitableUTxO {totalDeposit = Coin 80747850}
```

Which should be clearer for the future.

ch1bo commented on hydra-node/src/Hydra/Chain/ScriptRegistry.hs:
```diff
@@ -105,54 +102,56 @@ publishHydraScripts networkId socketPath sk = do
+
+-- | Builds a chain of script publishing transactions.
+-- Throws: PublishScriptException
 buildScriptPublishingTxs ::
```
@ffakenz @v0d1ch You might want to review this new implementation as you both have worked on it recently.
